src/swap.ts

```diff
@@ -0,0 +1,329 @@
+ *
+ * @returns Returns the manipulated array or string.
+ *
+ * @data_first
+ */
+export function swap<
+  T extends ReadonlyArray<unknown> | string,
+  K1 extends keyof T & number,
+  K2 extends keyof T & number
+>(data: T, key1: K1, key2: K2): Swap<T, K1, K2>;
+/**
+ * @param data the object to be manipulated
+ * @param key1 the first property key
+ * @param key2 the second property key
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @returns Returns the manipulated object.
+ *
+ * @example
+ *   swap({a: 1, b: 2, c: 3}, 'a', 'b') // => {a: 2, b: 1, c: 3}
+ */
+export function swap<T extends object, K1 extends keyof T, K2 extends keyof T>(
+  data: T,
+  key1: K1,
+  key2: K2
+): ObjectSwap<T, K1, K2>;
+
+/**
+ * @param key1 the first index
+ * @param key2 the second index
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @example
+ *   swap(0, 1)(['a', 'b', 'c']) // => ['b', 'a', 'c']
+ *   swap(0, -1)('abc') // => 'cba'
+ *
+ * @category Array
+ * @returns Returns the manipulated array or string.
+ * @data_last
+ */
+export function swap<K1 extends number, K2 extends number>(
+  key1: K1,
+  key2: K2
+): <T extends ReadonlyArray<unknown> | string>(data: T) => Swap<T, K1, K2>;
+
+/**
+ * Swaps the positions of two properties in an object based on the provided keys.
+ * @param key1 the first property key
+ * @param key2 the second property key
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @example
+ *   swap('a', 'b')({a: 1, b: 2, c: 3}) // => {a: 2, b: 1, c: 3}
+ *
+ * @returns Returns the manipulated object.
+ */
+export function swap<K1 extends PropertyKey, K2 extends PropertyKey>(
+  key1: K1,
+  key2: K2
+): <T extends { [P in K1 | K2]: any }>(data: T) => ObjectSwap<T, K1, K2>;
+
+/**
+ * Swaps the positions of two elements in a list, string, or object based on the provided keys.
+ *
+ * @returns A curried function that expects the item to be manipulated.
+ */
+export function swap() {
+  return purry(_swap, arguments);
+}
+
+function _swap(item: any, key1: any, key2: any): any {
+  if (isArray(item)) {
+    return _swapList(item, key1, key2);
+  } else if (isString(item)) {
+    return _swapString(item, key1, key2);
+  } else {
+    return _swapObject(item, key1, key2);
+  }
+}
+
+function _swapList<T>(
+  item: ReadonlyArray<T>,
+  index1: number,
+  index2: number
+): Array<T> {
+  const length = item.length;
+  let result = item.slice();
+
+  if (isNaN(index1) || isNaN(index2)) {
+    return result;
+  }
+
+  const positiveIndexA = index1 < 0 ? length + index1 : index1;
+  const positiveIndexB = index2 < 0 ? length + index2 : index2;
+
+  const positiveMin = Math.min(positiveIndexA, positiveIndexB);
+  const positiveMax = Math.max(positiveIndexA, positiveIndexB);
+
+  if (positiveIndexA < 0 || positiveIndexA > length) {
+    return result;
+  }
+  if (positiveIndexB < 0 || positiveIndexB > length) {
+    return result;
+  }
+  if (positiveIndexA === positiveIndexB) {
+    return result;
+  }
```

Comment:
I don't think we need to optimize for this very specific case, the swap logic is sound without this.
```suggestion
```